tpu-provisioner/internal/controller/creation_controller.go

```diff
@@ -96,5 +98,34 @@ func (r *CreationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 func (r *CreationReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&corev1.Pod{}).
+		WithEventFilter(predicate.NewPredicateFuncs(func(object client.Object) bool {
+			// Only reconcile JobSet leader pods which have not been scheduled and
+			// have not been marked for deletion.
+			pod, ok := object.(*corev1.Pod)
+			return ok && podPartOfJobSet(pod) && isLeaderPod(pod) && !podScheduled(pod) && !podDeleted(pod)
```

Comment:
Can you remove the similar checks from the Reconcile() func and move all to here and add something into the integration test that checks it is still working?

Reply:
Consolidated pod filtering logic in the event filter and updated integration test with more cases. I also refactored the integration test structure to be more organized and allow us to easily add more cases in the future.